staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -31,15 +31,59 @@ import (
 	"k8s.io/klog"
 )
 
-// SharedInformer has a shared data cache and is capable of distributing notifications for changes
-// to the cache to multiple listeners who registered via AddEventHandler. If you use this, there is
-// one behavior change compared to a standard Informer.  When you receive a notification, the cache
-// will be AT LEAST as fresh as the notification, but it MAY be more fresh.  You should NOT depend
-// on the contents of the cache exactly matching the notification you've received in handler
-// functions.  If there was a create, followed by a delete, the cache may NOT have your item.  This
-// has advantages over the broadcaster since it allows us to share a common cache across many
-// controllers. Extending the broadcaster would have required us keep duplicate caches for each
-// watch.
+// SharedInformer provides eventually consistent linkage of its
+// clients to the authoritative state of a given collection of
+// objects.  An object is identified by its API group, kind/resource,
+// namespace, and name.  One SharedInfomer provides linkage to objects
+// of a particular API group and kind/resource, and may be further
+// restricted to one namespace and/or by label selector and/or field
+// selector.
+//
+// The authoritative state of an object is what apiservers provide
+// access to, and an object goes through a strict sequence of states.
+// A state is either "absent" or present with a ResourceVersion and
+// other appropriate content.
+//
+// A SharedInformer maintains a local cache, exposed by Store(), of
+// the state of each relevant object.  This cache is eventually
+// consistent with the authoritative state.  This means that, unless
+// prevented by persistent communication problems, if ever a
+// particular object ID X is authoritatively associated with a state S
+// then for every SharedInformer I whose collection includes (X, S)
+// eventually either (1) I's cache associates X with S or a later
+// state of X or (2) I is stopped.  To be formally complete, we say
+// that the absent state meets any restriction by label selector or
+// field selector.
```

Comment:
Possibly worth including, I'm sure you can come up with a better wording:

The watch semantics also produce another property; since add/update/delete is processed in serial and the List() method is thread-safe (i.e., blocks processing updates), it follows that each list returned by the cache also corresponds to a complete list of objects that was actually stored by the apiserver at a point of time. (One can perhaps imagine alternative watch systems which don't offer this property, where the list as a whole isn't valid because some objects are further ahead in time than other objects.)

This property is only valid within a single group/version/kind. E.g. in many installations Events are stored in a different etcd cluster than, say, Pods. And even if they weren't, there's no way to watch multiple types in the same watch, and there's no way to guarantee that two watches stay in sync--are always at the same point in time--so this property can't be had across disparate GVKs.

Reply:
I have noticed that SharedInformer promises less than a single LIST or WATCH, since it is synthesized from a complex combination of those.  If I understand correctly, a SharedInformer does not promise any particular coherence between objects of the same GVK.  The only place where the coherence of a LIST result shows up is in the definition of what "HasSynced" means (which I am about to try to write down).

Reply:
I also do not want to see LIST or WATCH defined in a way that disallows sharding the storage of a single GVK.

Reply:
The last point is fair. We have not implemented any such sharding (I know you have mentioned it before) yet, but it is good not to tie our hands. I can't think of a specific controller which depends on this property, although we may have some.

SharedInformer would have to be doing something very weird to *not* get the property I mention (once HasSynced() returns true, anyway), since we do not actually do any sharding today.

Reply:
Oh, I see, you are talking about behavior of the Store.  Yes, that is something that could be added.

Reply:
> The watch semantics also produce another property; since add/update/delete is processed in serial and the List() method is thread-safe (i.e., blocks processing updates), it follows that each list returned by the cache also corresponds to a complete list of objects that was actually stored by the apiserver at a point of time.

IIUC, this is not true for a single GVK. A reflector writes to its store adds/updates/deletes in serial order, but the reflector's store is not the store that answers SharedIndexInformer lookups/list. A reflector's store is a  [DeltaFIFO](https://github.com/kubernetes/client-go/blob/master/tools/cache/delta_fifo.go#L96) which mainly consists of a [FIFO queue of object IDs](https://github.com/kubernetes/client-go/blob/master/tools/cache/delta_fifo.go#L105) (i.e. names and namespaced names), and a [map from object ID to a list of deltas](https://github.com/kubernetes/client-go/blob/master/tools/cache/delta_fifo.go#L104) (roughly speaking, delta = notification). A SharedIndexInformer runs a loop which serially pops object's IDs from the DeltaFIFO's queue and uses the map from object IDs to deltas to retrieve the object ID's deltas ([here](https://github.com/kubernetes/client-go/blob/master/tools/cache/controller.go#L152-L153)'s the loop, [here](https://github.com/kubernetes/client-go/blob/master/tools/cache/delta_fifo.go#L411)'s the Pop function). The catch is that if a new delta _D_ for object ID _X_ is "added" to the DeltaFIFO and [the DeltaFIFO already stores _X_ in the queue and there are older _X_'s deltas yet to be processed in the map from object ID to deltas, _D_ is added to the list of deltas for _X_ and _X_ is not added again to the queue](https://github.com/kubernetes/client-go/blob/master/tools/cache/delta_fifo.go#L324-L327). When the SharedIndexInformer pops _X_ from the DeltaFIFO's queue, [it applies to the cache all deltas for _X_  in a thread-safe way before processing any other object ID's deltas](https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go#L453-L475). Follows an example which leads to a violation of coherence of List() results. An object (_X_, _Sx_) where _X_ is the namespaced name and _Sx_ the state (i.e. spec, status, etc...) is created, then an object (_Y_, _Sy_) is created. After the reflector sees the corresponding notifications, the DeltaFIFO's queue will contain _X_ and _Y_ (assuming _X_ was not popped before _Y_'s addition) with _X_ destined to be popped first; the map from object IDs to deltas will associate _X_ to the list of deltas [(Create,_Sx_)] and _Y_ to the list of deltas [(Create,_Sy_)]. Assume that before _X_ is popped the reflector is notified of an update of _X_ to _Sx'_. The DeltaFIFO's queue will be unchanged, while the map from object ID to deltas now associates _X_ to the list of deltas [(Create,_Sx_);(Update,_Sx'_)]. When _X_ is popped the SharedIndexInformer modifies its cache by applying all _X_'s deltas to it, the final result is that the informer's cache contains (_X_, _Sx'_) and does not contain _Y_, this collection state was never stored by the API server, but is returned by a SharedIndexInformer List() invoked before _Y_ is popped from the DeltaFIFO.

In the aforementioned example, the problem is that deltas for different objects are applied to an informer's cache out of order. Even if deltas for WATCH notifications were applied to informer's caches in order, consistency of SharedIndexInformer.List() results can still be violated. Assume deltas for WATCH notifications are applied in order to SharedIndexInformer caches, also assume a SharedIndexInformer has been running for some time: the first API server list results have been loaded, `HasSynced()` returns true. Now assume a partition occurs, and when it disappears the reflector needs to do a re-list. The reflector adds deltas for all the objects in the LIST results to the DeltaFIFO in a thread-safe way (while holding the DeltaFIFO's mutex). But each object ID's list of deltas is popped and applied to the informer's cache separately: the informer's cache mutex is released between writing of two different object ID's deltas to the cache. The reflector adds object IDs to the DeltaFIFO in the order in which they appear in the LIST results, which is not "oldest to newest". If during the partition two new objects _X_ and _Y_ are created with _X_ being created first, they might be added to the DeltaFIFO in the opposite order; that's the same order with which they'll appear in the informer's cache. If _Y_ is added first and SharedIndexInformer.List() is invoked before _X_ is added to the informer's cache, a collection state that was never stored by the API server is returned.

Reply:
@MikeSpreitzer @lavalamp

Reply:
OK, the grouping of one object ID's deltas destroys the property that @lavalamp suggested and I documented; the comment needs to be updated.

Regarding the second paragraph of the preceding Comment, I thought that when I last looked at the code there was one mutex that was held throughout the insertion of LIST results and this excluded lookups by informer clients.

Reply:
Strictly speaking, I think none of the paragraphs you quoted need to be touched.
IIUC, consistency between cache and authoritative state modifications order is not necessary from the definition of “collection state” given in the first paragraph you quoted. The second paragraph you quoted makes sense even without ordering of writes: it means that an informer’s cache is not modified while populating list results, which is true and still relevant, as it guarantees that list results will never include two objects with the same ID. Regardless, it might be desirable to clarify the comments to clearly state that there's no ordering.

As for the breaking of ordering by relists, no, lookups from informer clients are not excluded by the mutex. However, that alone does not break ordering. If the API server populates list results in the order in which objects are created (I think not but I'm not 100% confident), current relists preserve ordering of informer cache modifications.

Reply:
I can find no documentation promising any particular ordering in list results.

Experimentally, I do not find list results to be ordered by increasing resource version.  I tried an HTTP GET of both namespaces and of pods.

Reply:
I am tired of trying to understand shared informers based on missing/wrong comments on the subsidiary abstractions used in the implementation.  I have started #86774 to address this.

Reply:
OK, I have made enough progress on #86774 to respond here.  @matte21 , your earliest remarks are key.  The shared informer's local cache (Indexer) is a distinct data structure from the DeltaFIFO, and what's done atomically is the Pop (and included processing) of a single Deltas from the fifo.  Note also that the Deleteds created by Replace follow Syncs.  That means that even if the LIST returned objects in resourceVersion order, and the fifo would not necessarily have keys in ResourceVersion order.

Thus the statement at https://github.com/kubernetes/kubernetes/blob/822a2de262372db4fb11aa4d69389c0cdbd869aa/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L117-L120 is not correct and should be deleted.

The definition at https://github.com/kubernetes/kubernetes/blob/822a2de262372db4fb11aa4d69389c0cdbd869aa/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L49-L56 makes sense but nobody will care; after deleting the incorrect statement, nothing will refer to the concept of "collection state".

Reply:
Reason (2), that is:
> some keys might already be in the fifo at the start of a Replace

and the following statement

> one key's Deltas can contain multiple Delta objects

are related, right?